ioSPI/fourier.py

```diff
@@ -0,0 +1,237 @@
+    Equivalent to fft shifting
+    https://dsp.stackexchange.com/questions/9039/centering-zero-frequency-for-discrete-fourier-transform
+
+    Parameters
+    ----------
+    arr3d_shape : tuple (N_fft_pix,N_fft_pix,N_fft_pix)
+        Input shape tuple.
+
+    Returns
+    -------
+    neg_pos_3d : numpy.ndarray, shape (N_fft_pix,N_fft_pix,N_fft_pix)
+        Checkerboard array.
+    """
+    R1, R2, R3 = arr3d_shape
+    neg_pos_3d = np.ones(arr3d_shape)
+    for r1 in range(R1):
+        for r2 in range(R2):
+            for r3 in range(R3):
+                if (r1 + r2 + r3) % 2:
+                    neg_pos_3d[r1, r2, r3] *= -1
+    return neg_pos_3d
+
+
+def fft3d(
+    arr3d,
+    mode,
+    neg_pos_3d=None,
+    numpy_fft=pyfftw.interfaces.numpy_fft,
+    only_real=False,
+):
+    """3D FFT
+
+    Wraps fft2d
+    Fast fourier transform. We apply an alterating +1/-1 multiplicative
+    before we go to/from Fourier space.
+    Later we apply this again to the transform.
+    The checkerboard pattern is applied instead of fft shifting
+    to have the dc component in centre of image
+    (even number of pixels, one to right of centre).
+
+    Parameters
+    ----------
+    arr3d : numpy.ndarray, shape (N,N,N)
+        Input array.
+    mode : str
+        Forward or reverse
+    neg_pos_3d : numpy.ndarray, shape (N,N,N)
+        Optional precomputed array. If None (default) passed,
+        computes on the fly.
+    numpy_fft : func
+        Function backend for performing the fft.
+    only_real : bool
+        If True, return only real part of transform.
+
+    Returns
+    -------
+    arr3d_f : numpy.ndarray, shape (N,N,N)
+        Output array.
+    """
+    # compute on the fly if not precomputed
+    if neg_pos_3d is None:
+        neg_pos_3d = make_neg_pos_3d(arr3d.shape)
+        # extra multiplication for mod 4
+        if arr3d.shape[0] % 4 != 0:
+            neg_pos_3d *= -1
+
+    if mode == "forward":
+        arr3d_f = numpy_fft.fftn(neg_pos_3d * arr3d)
+        arr3d_f /= np.sqrt(np.prod(arr3d_f.shape))
+    elif mode == "inverse":
+        arr3d_f = numpy_fft.ifftn(neg_pos_3d * arr3d)
+        arr3d_f *= np.sqrt(np.prod(arr3d_f.shape))
+
+    if only_real:
+        arr3d_f = arr3d_f.real
+
+    arr3d_f *= neg_pos_3d
+    return arr3d_f
+
+
+def fft2d(
+    arr2d,
+    mode,
+    numpy_fft=pyfftw.interfaces.numpy_fft,
+    only_real=False,
+    batch=False,
+):
+    """2D FFT
+
+    Fast fourier transform. We apply an alterating +1/-1 multiplicative
+    before we go to/from Fourier space.
+    Later we apply this again to the transform.
+    The checkerboard pattern is applied instead of fft shifting
+    to have the dc component in centre of image
+    (even number of pixels, one to right of centre).
+
+    Parameters
+    ----------
+    arr2d : numpy.ndarray, shape (N,N) or (num_exposures,N,N)
+        Input array.
+    mode : str
+        Forward or reverse
```

Comment:
Missing . at the end of the parater description

You can write:

`mode : str, {"forward", "reverse"}`
to indicate the possibility for this param.

Reply:
Please fix.

Reply:
Done